File: tests/wasm/test_macd.js
/**
 * WASM binding tests for MACD indicator.
 * These tests mirror the Rust unit tests to ensure WASM bindings work correctly.
 */
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { 
    loadTestData, 
    assertArrayClose, 
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS 
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {
    
    try {
        const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
        const importPath = process.platform === 'win32' 
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);
        
    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }
    
    testData = loadTestData();
});

test('MACD partial params', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.macd_js(close, 12, 26, 9, "ema");
    assert.strictEqual(result.values.length, close.length * 3); 
    assert.strictEqual(result.rows, 3);
    assert.strictEqual(result.cols, close.length);
});

test('MACD accuracy', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.macd_js(close, 12, 26, 9, "ema");
    
    
    const macd = result.values.slice(0, close.length);
    const signal = result.values.slice(close.length, close.length * 2);
    const hist = result.values.slice(close.length * 2);
    
    
    const expected_macd = [
        -629.8674025082801,
        -600.2986584356258,
        -581.6188884820076,
        -551.1020443476082,
        -560.798510688488,
    ];
    const expected_signal = [
        -721.9744591891067,
        -697.6392990384105,
        -674.4352169271299,
        -649.7685824112256,
        -631.9745680666781,
    ];
    const expected_hist = [
        92.10705668082664,
        97.34064060278467,
        92.81632844512228,
        98.6665380636174,
        71.17605737819008,
    ];
    
    const last5_macd = macd.slice(-5);
    const last5_signal = signal.slice(-5);
    const last5_hist = hist.slice(-5);
    
    
    assertArrayClose(
        last5_macd,
        expected_macd,
        1e-1,
        "MACD last 5 values mismatch"
    );
    assertArrayClose(
        last5_signal,
        expected_signal,
        1e-1,
        "Signal last 5 values mismatch"
    );
    assertArrayClose(
        last5_hist,
        expected_hist,
        1e-1,
        "Histogram last 5 values mismatch"
    );
});

test('MACD zero period', () => {
    
    const inputData = new Float64Array([10.0, 20.0, 30.0]);
    
    
    assert.throws(() => {
        wasm.macd_js(inputData, 0, 26, 9, "ema");
    }, /Invalid period|unreachable/);
});

test('MACD period exceeds length', () => {
    
    const data = new Float64Array([10.0, 20.0, 30.0]);
    
    assert.throws(() => {
        wasm.macd_js(data, 12, 26, 9, "ema");
    }, /Invalid period|unreachable/);
});

test('MACD very small dataset', () => {
    
    const data = new Float64Array([42.0]);
    
    assert.throws(() => {
        wasm.macd_js(data, 12, 26, 9, "ema");
    }, /Invalid period|Not enough valid data|unreachable/);
});

test('MACD empty input', () => {
    
    const empty = new Float64Array([]);
    
    assert.throws(() => {
        wasm.macd_js(empty, 12, 26, 9, "ema");
    }, /input data slice is empty|invalid period|unreachable/i);
});

test('MACD NaN handling', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.macd_js(close, 12, 26, 9, "ema");
    
    
    const macd = result.values.slice(0, close.length);
    const signal = result.values.slice(close.length, close.length * 2);
    const hist = result.values.slice(close.length * 2);
    
    
    if (close.length > 240) {
        for (let i = 240; i < close.length; i++) {
            assert(!isNaN(macd[i]), `Found unexpected NaN in MACD at index ${i}`);
            assert(!isNaN(signal[i]), `Found unexpected NaN in signal at index ${i}`);
            assert(!isNaN(hist[i]), `Found unexpected NaN in histogram at index ${i}`);
        }
    }
});

test('MACD all NaN input', () => {
    
    const allNaN = new Float64Array(100);
    allNaN.fill(NaN);
    
    assert.throws(() => {
        wasm.macd_js(allNaN, 12, 26, 9, "ema");
    }, /all values are nan/i);
});

test('MACD fast API (in-place)', () => {
    
    const close = new Float64Array(testData.close);
    const len = close.length;
    
    
    const in_ptr = wasm.macd_alloc(len);
    const macd_ptr = wasm.macd_alloc(len);
    const signal_ptr = wasm.macd_alloc(len);
    const hist_ptr = wasm.macd_alloc(len);
    
    try {
        
        const memory = wasm.__wasm.memory;
        const memView = new Float64Array(memory.buffer, in_ptr, len);
        memView.set(close);
        
        
        
        wasm.macd_into(
            in_ptr,
            macd_ptr,
            signal_ptr,
            hist_ptr,
            len,
            12, 26, 9, "ema"
        );
        
        
        
        const memory2 = wasm.__wasm.memory;
        const fullView = new Float64Array(memory2.buffer);
        const macd = fullView.subarray(macd_ptr / 8, macd_ptr / 8 + len);
        const signal = fullView.subarray(signal_ptr / 8, signal_ptr / 8 + len);
        const hist = fullView.subarray(hist_ptr / 8, hist_ptr / 8 + len);
        
        
        const macdCopy = Array.from(macd);
        const signalCopy = Array.from(signal);
        const histCopy = Array.from(hist);
        
        
        const safeResult = wasm.macd_js(close, 12, 26, 9, "ema");
        const safe_macd = safeResult.values.slice(0, len);
        const safe_signal = safeResult.values.slice(len, len * 2);
        const safe_hist = safeResult.values.slice(len * 2);
        
        
        assertArrayClose(macdCopy, safe_macd, 1e-10, "Fast API MACD mismatch");
        assertArrayClose(signalCopy, safe_signal, 1e-10, "Fast API signal mismatch");
        assertArrayClose(histCopy, safe_hist, 1e-10, "Fast API histogram mismatch");
    } finally {
        
        wasm.macd_free(in_ptr, len);
        wasm.macd_free(macd_ptr, len);
        wasm.macd_free(signal_ptr, len);
        wasm.macd_free(hist_ptr, len);
    }
});

test('MACD fast API aliasing detection', () => {
    
    const close = new Float64Array(testData.close);
    const len = close.length;
    
    
    const buffer_ptr = wasm.macd_alloc(len);
    const signal_ptr = wasm.macd_alloc(len);
    const hist_ptr = wasm.macd_alloc(len);
    
    try {
        
        const memory = wasm.__wasm.memory;
        const buffer = new Float64Array(memory.buffer, buffer_ptr, len);
        buffer.set(close);
        
        
        
        wasm.macd_into(
            buffer_ptr,
            buffer_ptr, 
            signal_ptr,
            hist_ptr,
            len,
            12, 26, 9, "ema"
        );
        
        
        
        const memory2 = wasm.__wasm.memory;
        const fullView = new Float64Array(memory2.buffer);
        const buffer_after = fullView.subarray(buffer_ptr / 8, buffer_ptr / 8 + len);
        
        
        const bufferCopy = Array.from(buffer_after);
        
        
        const safeResult = wasm.macd_js(close, 12, 26, 9, "ema");
        const safe_macd = safeResult.values.slice(0, len);
        
        assertArrayClose(bufferCopy, safe_macd, 1e-10, "Aliasing handling failed");
    } finally {
        
        wasm.macd_free(buffer_ptr, len);
        wasm.macd_free(signal_ptr, len);
        wasm.macd_free(hist_ptr, len);
    }
});

test('MACD batch single parameter set', () => {
    
    const close = new Float64Array(testData.close);
    
    const batchResult = wasm.macd_batch(close, {
        fast_period_range: [12, 12, 0],
        slow_period_range: [26, 26, 0],
        signal_period_range: [9, 9, 0],
        ma_type: "ema"
    });
    
    
    assert.strictEqual(batchResult.rows, 1);
    assert.strictEqual(batchResult.cols, close.length);
    assert.strictEqual(batchResult.values.length, 3 * close.length); 
    
    
    const batch_macd = batchResult.values.slice(0, close.length);
    const batch_signal = batchResult.values.slice(close.length, 2 * close.length);
    const batch_hist = batchResult.values.slice(2 * close.length);
    
    
    const singleResult = wasm.macd_js(close, 12, 26, 9, "ema");
    const single_macd = singleResult.values.slice(0, close.length);
    const single_signal = singleResult.values.slice(close.length, 2 * close.length);
    const single_hist = singleResult.values.slice(2 * close.length);
    
    assertArrayClose(
        batch_macd,
        single_macd,
        1e-10,
        "Batch vs single MACD mismatch"
    );
    assertArrayClose(
        batch_signal,
        single_signal,
        1e-10,
        "Batch vs single signal mismatch"
    );
    assertArrayClose(
        batch_hist,
        single_hist,
        1e-10,
        "Batch vs single histogram mismatch"
    );
});

test('MACD batch multiple parameters', () => {
    
    const close = new Float64Array(testData.close.slice(0, 50)); 
    
    
    const batchResult = wasm.macd_batch(close, {
        fast_period_range: [10, 14, 2],   
        slow_period_range: [24, 28, 2],   
        signal_period_range: [8, 10, 1],  
        ma_type: "ema"
    });
    
    
    assert.strictEqual(batchResult.rows, 27);
    assert.strictEqual(batchResult.cols, 50);
    assert.strictEqual(batchResult.values.length, 3 * 27 * 50); 
    assert.strictEqual(batchResult.fast_periods.length, 27);
    assert.strictEqual(batchResult.slow_periods.length, 27);
    assert.strictEqual(batchResult.signal_periods.length, 27);
    
    
    
    const macd_block_size = 27 * 50;
    const firstMacd = batchResult.values.slice(0, 50); 
    
    const singleResult = wasm.macd_js(close, 10, 24, 8, "ema");
    const single_macd = singleResult.values.slice(0, 50);
    
    assertArrayClose(
        firstMacd,
        single_macd,
        1e-10,
        "First batch row should match single calculation"
    );
});

test('MACD unknown MA type', () => {
    
    const data = new Float64Array([10.0, 20.0, 30.0, 40.0, 50.0]);
    
    assert.throws(() => {
        wasm.macd_js(data, 2, 3, 2, "unknown_ma");
    }, /Unknown MA type/);
});

test('MACD warmup periods', () => {
    
    const close = new Float64Array(testData.close);
    const fastPeriod = 12;
    const slowPeriod = 26;
    const signalPeriod = 9;
    
    const result = wasm.macd_js(close, fastPeriod, slowPeriod, signalPeriod, "ema");
    
    
    const macd = result.values.slice(0, close.length);
    const signal = result.values.slice(close.length, close.length * 2);
    const hist = result.values.slice(close.length * 2);
    
    
    const macdWarmup = slowPeriod - 1;
    for (let i = 0; i < macdWarmup; i++) {
        assert(isNaN(macd[i]), `Expected NaN at MACD index ${i} during warmup`);
    }
    
    
    const signalWarmup = slowPeriod + signalPeriod - 2;
    for (let i = 0; i < signalWarmup; i++) {
        assert(isNaN(signal[i]), `Expected NaN at signal index ${i} during warmup`);
        assert(isNaN(hist[i]), `Expected NaN at histogram index ${i} during warmup`);
    }
    
    
    assert(!isNaN(macd[macdWarmup]), `Unexpected NaN at MACD index ${macdWarmup} after warmup`);
    assert(!isNaN(signal[signalWarmup]), `Unexpected NaN at signal index ${signalWarmup} after warmup`);
    assert(!isNaN(hist[signalWarmup]), `Unexpected NaN at histogram index ${signalWarmup} after warmup`);
});

test('MACD different MA types', () => {
    
    const close = new Float64Array(testData.close.slice(0, 100)); 
    
    
    const resultEMA = wasm.macd_js(close, 12, 26, 9, "ema");
    const macdEMA = resultEMA.values.slice(0, close.length);
    assert.strictEqual(macdEMA.length, close.length);
    
    
    const resultSMA = wasm.macd_js(close, 12, 26, 9, "sma");
    const macdSMA = resultSMA.values.slice(0, close.length);
    assert.strictEqual(macdSMA.length, close.length);
    
    
    const resultWMA = wasm.macd_js(close, 12, 26, 9, "wma");
    const macdWMA = resultWMA.values.slice(0, close.length);
    assert.strictEqual(macdWMA.length, close.length);
    
    
    let emaVsSmaMatch = true;
    let emaVsWmaMatch = true;
    for (let i = 50; i < close.length; i++) {
        if (Math.abs(macdEMA[i] - macdSMA[i]) > 1e-5) emaVsSmaMatch = false;
        if (Math.abs(macdEMA[i] - macdWMA[i]) > 1e-5) emaVsWmaMatch = false;
    }
    assert(!emaVsSmaMatch, "EMA and SMA should produce different results");
    assert(!emaVsWmaMatch, "EMA and WMA should produce different results");
});

test('MACD batch edge cases', () => {
    
    const close = new Float64Array(testData.close.slice(0, 50));
    
    
    const largeBatch = wasm.macd_batch(close, {
        fast_period_range: [12, 14, 10], 
        slow_period_range: [26, 26, 0],
        signal_period_range: [9, 9, 0],
        ma_type: "ema"
    });
    
    
    assert.strictEqual(largeBatch.rows, 1);
    assert.strictEqual(largeBatch.fast_periods.length, 1);
    assert.strictEqual(largeBatch.fast_periods[0], 12);
    
    
    const batchResult = wasm.macd_batch(close, {
        fast_period_range: [10, 14, 2],   
        slow_period_range: [24, 28, 2],   
        signal_period_range: [8, 8, 1],   
        ma_type: "ema"
    });
    
    
    assert.strictEqual(batchResult.rows, 9);
    
    
    
    const firstRowMacd = batchResult.values.slice(0, 50);
    const firstRowSignal = batchResult.values.slice(9 * 50, 9 * 50 + 50);
    const firstRowHist = batchResult.values.slice(2 * 9 * 50, 2 * 9 * 50 + 50);
    
    
    for (let i = 0; i < 23; i++) {
        assert(isNaN(firstRowMacd[i]), `Expected NaN in batch MACD at index ${i}`);
    }
    
    
    for (let i = 0; i < 30; i++) {
        assert(isNaN(firstRowSignal[i]), `Expected NaN in batch signal at index ${i}`);
        assert(isNaN(firstRowHist[i]), `Expected NaN in batch histogram at index ${i}`);
    }
    
    
    assert.throws(() => {
        wasm.macd_batch(new Float64Array([]), {
            fast_period_range: [12, 12, 0],
            slow_period_range: [26, 26, 0],
            signal_period_range: [9, 9, 0],
            ma_type: "ema"
        });
    }, /All values are NaN|Input data slice is empty/);
});

test('MACD memory allocation/deallocation', () => {
    
    const len = 1000;
    
    
    const ptr = wasm.macd_alloc(len);
    assert(ptr !== 0, "Allocation should return non-null pointer");
    
    
    assert.doesNotThrow(() => {
        wasm.macd_free(ptr, len);
    });
    
    
    assert.doesNotThrow(() => {
        wasm.macd_free(0, len);
    });
});


test.after(() => {
    if (process.env.RUN_RUST_COMPARISON) {
        compareWithRust('macd', testData);
    }
});

File: tests/python/test_macd.py
"""
Python binding tests for MACD indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:
    
    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close
from rust_comparison import compare_with_rust


class TestMacd:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()
    
    def test_macd_partial_params(self, test_data):
        """Test MACD with default parameters - mirrors check_macd_partial_params"""
        close = test_data['close']
        
        
        macd, signal, hist = ta_indicators.macd(close, 12, 26, 9, "ema")
        assert len(macd) == len(close)
        assert len(signal) == len(close)
        assert len(hist) == len(close)
    
    def test_macd_accuracy(self, test_data):
        """Test MACD matches expected values from Rust tests - mirrors check_macd_accuracy"""
        close = test_data['close']
        
        macd, signal, hist = ta_indicators.macd(close, 12, 26, 9, "ema")
        
        assert len(macd) == len(close)
        assert len(signal) == len(close)
        assert len(hist) == len(close)
        
        
        expected_macd = [
            -629.8674025082801,
            -600.2986584356258,
            -581.6188884820076,
            -551.1020443476082,
            -560.798510688488,
        ]
        expected_signal = [
            -721.9744591891067,
            -697.6392990384105,
            -674.4352169271299,
            -649.7685824112256,
            -631.9745680666781,
        ]
        expected_hist = [
            92.10705668082664,
            97.34064060278467,
            92.81632844512228,
            98.6665380636174,
            71.17605737819008,
        ]
        
        
        assert_close(
            macd[-5:], 
            expected_macd,
            rtol=1e-1,
            msg="MACD last 5 values mismatch"
        )
        assert_close(
            signal[-5:], 
            expected_signal,
            rtol=1e-1,
            msg="Signal last 5 values mismatch"
        )
        assert_close(
            hist[-5:], 
            expected_hist,
            rtol=1e-1,
            msg="Histogram last 5 values mismatch"
        )
    
    def test_macd_zero_period(self):
        """Test MACD fails with zero period - mirrors check_macd_zero_period"""
        input_data = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(ValueError, match="Invalid"):
            ta_indicators.macd(input_data, 0, 26, 9, "ema")
    
    def test_macd_period_exceeds_length(self):
        """Test MACD fails when period exceeds data length - mirrors check_macd_period_exceeds_length"""
        data = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(ValueError, match="Invalid"):
            ta_indicators.macd(data, 12, 26, 9, "ema")
    
    def test_macd_very_small_dataset(self):
        """Test MACD fails with insufficient data - mirrors check_macd_very_small_dataset"""
        data = np.array([42.0])
        
        with pytest.raises(ValueError, match="Invalid|Not enough"):
            ta_indicators.macd(data, 12, 26, 9, "ema")
    
    def test_macd_nan_handling(self, test_data):
        """Test MACD NaN handling - mirrors check_macd_nan_handling"""
        close = test_data['close']
        
        macd, signal, hist = ta_indicators.macd(close, 12, 26, 9, "ema")
        n = len(macd)
        
        
        if n > 240:
            for i in range(240, n):
                assert not np.isnan(macd[i])
                assert not np.isnan(signal[i])
                assert not np.isnan(hist[i])
    
    def test_macd_streaming(self):
        """Test MACD streaming functionality"""
        stream = ta_indicators.MacdStream(12, 26, 9, "ema")
        
        
        values = [10.0, 20.0, 30.0, 40.0, 50.0, 60.0, 70.0, 80.0, 90.0, 100.0]
        result = None
        
        for i, value in enumerate(values):
            result = stream.update(value)
            
            
        
        
        assert result is None
        
        
        for i in range(100):
            result = stream.update(50.0 + i)
        
        
        assert result is not None
        assert isinstance(result, tuple)
        assert len(result) == 3  
    
    def test_macd_batch(self, test_data):
        """Test MACD batch processing"""
        close = test_data['close'][:100]  
        
        
        result = ta_indicators.macd_batch(
            close,
            (12, 12, 1),  
            (26, 26, 1),  
            (9, 9, 1),    
            "ema"
        )
        
        assert 'macd' in result
        assert 'signal' in result
        assert 'hist' in result
        assert 'fast_periods' in result
        assert 'slow_periods' in result
        assert 'signal_periods' in result
        
        
        assert result['macd'].shape == (1, 100)
        assert result['signal'].shape == (1, 100)
        assert result['hist'].shape == (1, 100)
        
        
        single_macd, single_signal, single_hist = ta_indicators.macd(close, 12, 26, 9, "ema")
        
        assert_close(
            result['macd'][0], 
            single_macd, 
            rtol=1e-10, 
            msg='Batch MACD should match single calculation'
        )
        assert_close(
            result['signal'][0], 
            single_signal, 
            rtol=1e-10, 
            msg='Batch signal should match single calculation'
        )
        assert_close(
            result['hist'][0], 
            single_hist, 
            rtol=1e-10, 
            msg='Batch hist should match single calculation'
        )
    
    def test_macd_batch_multiple_params(self, test_data):
        """Test MACD batch with multiple parameter combinations"""
        close = test_data['close'][:50]  
        
        
        result = ta_indicators.macd_batch(
            close,
            (10, 14, 2),   
            (24, 28, 2),   
            (8, 10, 1),    
            "ema"
        )
        
        
        assert result['macd'].shape == (27, 50)
        assert result['signal'].shape == (27, 50)
        assert result['hist'].shape == (27, 50)
        assert len(result['fast_periods']) == 27
        assert len(result['slow_periods']) == 27
        assert len(result['signal_periods']) == 27
        
        
        first_macd, first_signal, first_hist = ta_indicators.macd(close, 10, 24, 8, "ema")
        
        assert_close(
            result['macd'][0], 
            first_macd, 
            rtol=1e-10, 
            msg='First batch row should match single calculation'
        )
    
    def test_macd_warmup_periods(self, test_data):
        """Test MACD warmup periods match expected values"""
        close = test_data['close']
        fast_period = 12
        slow_period = 26
        signal_period = 9
        
        macd, signal, hist = ta_indicators.macd(close, fast_period, slow_period, signal_period, "ema")
        
        
        macd_warmup = slow_period - 1
        for i in range(macd_warmup):
            assert np.isnan(macd[i]), f"Expected NaN at MACD index {i} during warmup"
        
        
        signal_warmup = slow_period + signal_period - 2
        for i in range(signal_warmup):
            assert np.isnan(signal[i]), f"Expected NaN at signal index {i} during warmup"
            assert np.isnan(hist[i]), f"Expected NaN at histogram index {i} during warmup"
        
        
        assert not np.isnan(macd[macd_warmup]), f"Unexpected NaN at MACD index {macd_warmup} after warmup"
        assert not np.isnan(signal[signal_warmup]), f"Unexpected NaN at signal index {signal_warmup} after warmup"
        assert not np.isnan(hist[signal_warmup]), f"Unexpected NaN at histogram index {signal_warmup} after warmup"
    
    def test_macd_all_nan_input(self):
        """Test MACD with all NaN values"""
        all_nan = np.full(100, np.nan)
        
        with pytest.raises(ValueError, match="All values are NaN"):
            ta_indicators.macd(all_nan, 12, 26, 9, "ema")
    
    def test_macd_different_ma_types(self, test_data):
        """Test MACD with different moving average types"""
        close = test_data['close'][:100]  
        
        
        macd_ema, signal_ema, hist_ema = ta_indicators.macd(close, 12, 26, 9, "ema")
        assert len(macd_ema) == len(close)
        
        
        macd_sma, signal_sma, hist_sma = ta_indicators.macd(close, 12, 26, 9, "sma")
        assert len(macd_sma) == len(close)
        
        
        macd_wma, signal_wma, hist_wma = ta_indicators.macd(close, 12, 26, 9, "wma")
        assert len(macd_wma) == len(close)
        
        
        assert not np.allclose(macd_ema[50:], macd_sma[50:], rtol=1e-5), "EMA and SMA should produce different results"
        assert not np.allclose(macd_ema[50:], macd_wma[50:], rtol=1e-5), "EMA and WMA should produce different results"
        
        
        with pytest.raises(ValueError, match="Unknown MA type"):
            ta_indicators.macd(close, 12, 26, 9, "invalid_ma")
    
    def test_macd_batch_edge_cases(self, test_data):
        """Test MACD batch processing edge cases"""
        close = test_data['close'][:50]
        
        
        result = ta_indicators.macd_batch(
            close,
            (12, 14, 10),  
            (26, 26, 1),   
            (9, 9, 1),     
            "ema"
        )
        assert result['macd'].shape[0] == 1, "Should only have one parameter combination"
        
        
        empty = np.array([])
        with pytest.raises(ValueError, match="All values are NaN|Input data slice is empty"):
            ta_indicators.macd_batch(
                empty,
                (12, 12, 1),
                (26, 26, 1),
                (9, 9, 1),
                "ema"
            )
        
        
        result = ta_indicators.macd_batch(
            close,
            (10, 14, 2),   
            (24, 28, 2),   
            (8, 8, 1),     
            "ema"
        )
        
        
        assert result['macd'].shape[0] == 9
        
        
        first_row_macd = result['macd'][0]
        first_row_signal = result['signal'][0]
        first_row_hist = result['hist'][0]
        
        
        for i in range(23):
            assert np.isnan(first_row_macd[i]), f"Expected NaN in batch MACD at index {i}"
        
        
        for i in range(30):
            assert np.isnan(first_row_signal[i]), f"Expected NaN in batch signal at index {i}"
            assert np.isnan(first_row_hist[i]), f"Expected NaN in batch histogram at index {i}"
    
    def test_macd_kernel_parameter(self, test_data):
        """Test MACD with different kernel parameters"""
        close = test_data['close'][:100]
        
        
        macd_scalar, signal_scalar, hist_scalar = ta_indicators.macd(close, 12, 26, 9, "ema", "scalar")
        assert len(macd_scalar) == len(close)
        
        
        macd_auto, signal_auto, hist_auto = ta_indicators.macd(close, 12, 26, 9, "ema")
        assert len(macd_auto) == len(close)
        
        
        assert_close(macd_scalar, macd_auto, rtol=1e-12)
        assert_close(signal_scalar, signal_auto, rtol=1e-12)
        assert_close(hist_scalar, hist_auto, rtol=1e-12)


if __name__ == "__main__":
    pytest.main([__file__, "-v"])